packages/fxa-auth-server/lib/routes/emails.js

```diff
@@ -367,157 +358,143 @@ module.exports = (
+        if (type && type === 'secondary') {
+          await verifySecondaryEmail(account);
+          return {};
+        }
 
-                // Attempt to verify email token not associated with account
-                if (!isEmailVerification) {
-                  throw error.invalidVerificationCode();
-                }
-
-                // User is attempting to verify a secondary email that has already been verified.
-                // Silently succeed and don't send post verification email.
-                if (matchedEmail.isVerified) {
-                  log.info('account.verifyEmail.secondary.already-verified', {
-                    uid,
-                    code,
-                  });
-                  return P.resolve();
-                }
-
-                return db.verifyEmail(account, code).then(() => {
-                  log.info('account.verifyEmail.secondary.confirmed', {
-                    uid,
-                    code,
-                  });
-
-                  return mailer.sendPostVerifySecondaryEmail([], account, {
-                    acceptLanguage: request.app.acceptLanguage,
-                    secondaryEmail: matchedEmail.email,
-                    service,
-                    uid,
-                  });
-                });
+        const isAccountVerification = butil.buffersAreEqual(
+          code,
+          account.emailCode
+        );
+
+        let device;
+
+        try {
+          device = await db.deviceFromTokenVerificationId(uid, code);
+        } catch (err) {
+          if (err.errno !== error.ERRNO.DEVICE_UNKNOWN) {
+            log.error('Account.RecoveryEmailVerify', {
+              err,
+              uid,
+              code,
+            });
+          }
+        }
+
+        await accountAndTokenVerification(isAccountVerification, account);
+
+        if (device) {
+          const devices = request.app.devices;
+          const otherDevices = devices.filter(d => d.id !== device.id);
+          await push.notifyDeviceConnected(uid, otherDevices, device.name);
+        }
+
+        // If the account is already verified, the link may have been
+        // for sign-in confirmation or they may have been clicking a
+        // stale link. Silently succeed.
+        if (account.emailVerified) {
+          return {};
+        }
+
+        // Any matching code verifies the account
+        await signupUtils.verifyAccount(request, account, request.payload);
+
+        return {};
+
+        async function verifySecondaryEmail(account) {
+          let matchedEmail;
+          const emails = await db.accountEmails(uid);
+          const isEmailVerification = emails.some(email => {
+            if (email.emailCode && code === email.emailCode) {
+              matchedEmail = email;
+              log.info('account.verifyEmail.secondary.started', {
+                uid,
+                code,
               });
+              return true;
```

Comment:
This return value isn't actually used, is it needed?

Reply:
That's the return value for the `emails.some()` callback.